[Dart 3] Patterns (#4810)
dart-lang/site-www: src/language/patterns.md

---
title: Patterns
description: Summary of patterns in Dart.
---

{{site.alert.version-note}}
  Patterns require a [language version][] of at least 3.0.
{{site.alert.end}}

Patterns are a syntactic category in the Dart language, like statements and expressions.

> They represent the form of a value, not an actual value.

Reviewer: Maybe "shape of values"?

Reviewer 2: I may be wrong, but I think of a constant pattern as an "actual value"

Reviewer 3: I think of it as a constant pattern *has* a value for the constant that it tests, just like a relational pattern has a value for the right-hand side operand. But I don't think of patterns as being value-like in general.

The "set of values" and "actual values" feels like a confusing level of repetition. The only "set of values" the pattern represents are the actual ones it would match.

If you look up "pattern" in the dictionary, the two definitions that are closest are:

> * an arrangement or sequence regularly found in comparable objects. "the house had been built on the usual pattern"

> * a regular and intelligible form or sequence discernible in certain actions or situations. "a complicating factor is the change in working patterns"

I think, "regular", "comparable", and "discernable" are the key bits.

A pattern describes some commonality that a set of objects share and you can determine whether a given object has that commonality or not. A pattern is like a mold, gauge, or yardstick that you can hold an object up to and see if it fits.

So maybe, "A pattern describes a set of properties an object might have. A value can be *matched* against the pattern to see if it has those properties."

This page describes:
- What patterns do (match and destructure)
- Where patterns are allowed in Dart code
- Common use cases for patterns

To learn about the different types of patterns, visit the [pattern type reference][types].

## Matching

A pattern is always tested against a value to determine if the value has the form
you expect, or in other words, if the value _matches_ the pattern. 

What constitutes a match depends on the [pattern type][types].

> For example, matching a constant pattern means the value is equivalent to the pattern:

Reviewer: I don't know what "the value is equivalent to the pattern" means here.  What is this sentence trying to say?  Something like "For example, the following switch usess a constant pattern to test that `number` is `1`?  I'm not sure.

```dart
switch (number) {

>   case 1: ...        // Constant pattern 1 matches if number == 1

Reviewer: Technically, this tests `1 == number`, not sure if we want to be persnickety about it.

}
```

Many patterns make use of subpatterns (sometimes called outer and inner patterns,
respectively). Patterns match recursively on their subpatterns.
For example, the individual fields of any [collection-type][] pattern could be 
[variable patterns][variable] or [constant patterns][constant]:

```dart
switch (obj) {       // List pattern [a,b] matches obj first if obj
  case [a, b]: ...   // is a list wth two fields, then if its  
}                    // fields match the constant subpatterns a and b  
```

> Patterns can also _partially_ match, by using a [wildcard pattern][] as a place holder,

Reviewer: partially match is a somewhat odd term.  The pattern fully matches, it just "doesn't care" about certain parts of the value.  Maybe something like "Patterns can ignore pieces of the matched value that they don't care about by using a..."?

or a [rest element][] in the case of list patterns.

> If you need to ensure `null` values aren't silently treated as match failures,

Reviewer: This seems a little hard to situate without some context.  Maybe an example here?

use the [null-check pattern][] while matching.

## Destructuring

When an object and pattern match, the pattern can then access the object's data 
and extract it in parts. In other words, the pattern _destructures_ the object:

```dart
var numList = [1, 2, 3];
var [a, b, c] = numList;   // List pattern [a, b, c] destructures the list object
print(a + b + c);          // numList into individual new parts. Prints 6.
```

> You can use the [null-check pattern][] to check for `null` while destructuring. 

Reviewer: Well, this gets to my comment below.  You can use the null check pattern to destructure a value in a switch, but not in a destructuring assignment (which I think is more what this section is about).

Reviewer: Or maybe you don't indent this to just be about destructuring assignment?  But then it seems somewhat redundant with the previous section?  I guess I'm not clear on how these two sections fit together.

Reviewer: The `(int?, int?)` syntax there is a record type not a pattern.  The `p?` syntax meaning "nullable type" as a type and "check for null" is unfortunately very confusing - we just never found anything better.

Author: Oh, ok. I wonder if this warrants explaining somewhere, though I'm not sure where that explanation would go.

> All patterns attempt to match the objects they're compared against, but not all

Reviewer: I'm not sure about this paragraph.  What precisely do you want to the reader to take away here?  It's true that whether a pattern destructures or not depends on the type of pattern, not sure about the context.  Maybe there's some confusion around destructure in the sense of "destructuring assignment" vs "extract into parts"?  That is, the following pattern destructures (extracts) a list into parts, whether it is used is in a destructuring assignment, or in a switch:

```
  var [a, b] = someList; // Destructuring assignment: unconditionally extract the parts from something that must be a list.
  switch (someList) {
     case [var a, var b]: // Switch case: extract the parts from something that might be a list, if it is a list (and otherwise fall through to the next case).
  }
```

Reviewer: I think I somewhat misunderstood the point of this and the previous section, which perhaps partially reflects me being too close to the feature, and maybe reflects me reading to quickly, but might also suggest that other people might also go astray?  I think what you are trying to say with these sections is roughly "Hey, in general, there are two things going on in patterns.  First, given a value, you might want to check whether the value has a certain shape, or is a certain constant, or is equal to something else, or has a certain type.  And second, given a value, you might want a convenient declarative syntax to break it into a bunch of its constituent parts, possibly binding variables to some or all of those parts in the process.  The first we call matching, and the second we call destructuring.  Patterns let us to both at once. Depending on the context and the shape of the pattern, a pattern may match a value, destructure a value, or both.  Here are some examples to show you what we mean by matching, and here are some to show you what we mean by destructuring".  Is that a fair (rough summary)?  If so, I think these sections are actually pretty close, I just think that maybe a little bit of introductory text like the above to give the reader a roadmap of what's going on might avoid sending people like me down the wrong road (I was kind of reading this as talking more about refutable vs irrefutable contexts, whereas I think you mean something more like what I describe above).  Does that make sense?

Author: Yes that's it exactly. I added a new section header to summarize this, and then made the matching and destructuring sections subsections under that introduction

patterns destructure.
Whether a pattern destructures depends on the [type of pattern][types]
and the [context](#pattern-uses) in which it's being used.

## Pattern uses

Patterns integrate into several code constructs across the Dart language:

- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
- [for and for-in loops][for]
- [if-cases] and [switch-cases]
- [Collection literals][]

This section describes common use cases for matching and destructuring with patterns, 
combining various contexts, pattern types, and surrounding code constructs.

### Variable declaration

> Anywhere Dart allows regular, single-name local variable declaration, you can also

Reviewer 3: The "single-name" part throws me off a little. Maybe just: "In the body of a function where you can declare local variables, you can also use a _pattern variable declaration._"

Or do you intend to cover patterns in for and for-in statements/elements here too? (The variable declarations in those are also locals.)

use a _pattern variable declaration_. If the pattern matches the value on the right
of the declaration, it destructures the value, binding it to new local variables:

```dart

> var (a, [b, c]) = ("str", [1, 2]);   // Declares new variables a, b, and c.

Reviewer 4: nit:
```suggestion
// Declares new variables a, b, and c.
var (a, [b, c]) = ("str", [1, 2]);  
```

```

A pattern variable declaration must start with either `var` or `final`, followed
by a pattern. 

> Use the [null-assert pattern][] to eliminate `null` in variable declarations.

Reviewer: This should probably be phrased more carefully.  It eliminates `null` by throwing an exception, which makes this dangerous as blanket advice.

Reviewer 3: +1. I just wouldn't talk about `!` here.

### Variable assignment 

A _variable assignment pattern_ on the left side of an assignment destructures the
matched object and assigns the values to _existing_ variables,
instead of binding new ones. 

One convenient use case is swapping the values of two variables without needing
to declare a third temporary one:

```dart

> var (a, b) = ('left', 'right');
> (b, a) = (a, b);                   // Swap
> print('$a $b');                    // Prints "right left".

Reviewer 4: nit: for maintainability, don't align in-line comments manually:
```suggestion
var (a, b) = ('left', 'right');
(b, a) = (a, b); // Swap
print('$a $b'); // Prints "right left".
```

Reviewer 4: For inline comments, the convention I usually see is to add one space before `//`.

```

> ### Destrcuturing multiple returns

Reviewer: Destructuring (sp)

> ### Destrcuturing multiple returns

Reviewer 3: This section is an outlier because the other sections are describing distinct places in the language grammar where you can use a pattern. This section is more like a use case or idiom for patterns.

Maybe instead of having this as a separate section, you could just use this code as an example in the pattern variable declaration section sort of like how you use swapping as an example idiom under pattern assignment?

Taking a step back, maybe that's a smoother way to integrate the "places you can use a pattern" and "common idioms for patterns" together? For each place you use can use a pattern, the idiom then serves as a concrete example. Something like:

* Pattern variable declarations -> Destructuring multiple returns
* Pattern assignment -> Swapping variables
* Pattern variables in for and for-in -> The `MapEntry(...)` example of using an object pattern for destructuring
* Switch statements and expressions -> Algebraic datatype-style code with exhaustiveness checking over a sealed class's subtypes

Author: This is MUCH better, will do!

[Records][] allow aggregating and returning multiple values from a single function
call. Patterns add the ability to destructure a record's fields
directly into local variables, inline with the function call.

Instead of individually declaring new local variables for each record field,
like this:

```dart
main() {
  var info = userInfo(json);
  var name = info.$1;
  var age = info.$2;
}
```
You can call the function directly into a [variable declaration](#variable-declaration)
or [assigment pattern](#variable-assignment),

> with a [record subpattern][record] and fields for each value in the return:

Reviewer 4: The phrase "call a function into" doesn't parse for me. Maybe "You can destructure the fields of the record into local variables using a variable declaration pattern and a record pattern as its subpattern"

```dart
main() {
  var (name, age) = userInfo(json);
}
```

### Switch cases

> The `case` keyword is always followed by a pattern. _Case patterns_ are [refutable][];

Reviewer 3: This is true but doesn't also cover switch *expressions* where the pattern isn't preceded by `case`. (It was in early versions of the proposal, but we changed the syntax.)

Maybe say "Every case in a switch statement or expression contains a pattern."

they allow control flow to either match and destructure the object being switched
on, or continue execution if the object doesn't match. Case patterns can be any
kind of pattern:

- [**Constant patterns**][constant]: the object's value has to be equivalent to match.
  ```dart
  switch (number) {

>     case 1: ...          // Matches if number == 1  

Reviewer: Technically `1 == number`

  }
  ```
- [**List**][list], [**map**][map], or [**record**][record] **patterns**: the elements are destructured if matched.
  ```dart
  switch (obj) {
    case [a, b]: ...     // Matches if obj is a list with two elements  
  }
  ```  
- [**Variable patterns**][variable]: the matched value binds to a new variable.
  ```dart
  switch ((1, 2)) {

>     case (var a, var b): ...    // Matches if 1 and 2 can be assigned to a and b

Reviewer: This one is a bit odd.  As written, it always matches: that is "if 1 and 2 can be assigned to a and b" is always true.  Maybe change `(1, 2)` to `obj` and replace the comment with `// Matches if obj is a record with two fields, and if so assign the fields to a and b`?

  }
  ```  

> - [**Logical**][logical] or [**relational**][relational] **patterns**: the object matches if the comparison to

Reviewer: "The object matches if the comparison.." describes relational patterns only, so it reads a bit odd here when used to decribe "Logical or relational patterns".

a constant returns `true`.
  ```dart
  switch (code) {
    case >= first && <= last: ...    
    // Matches if the value of code is between the constant values of first and last
  }
  ```  
- [**Null-check**][check] or [**null-assert**][assert] **patterns**: the match is first on whether the object is not null, then on the value.
  ```dart
  switch (maybe) {
    case var s?: ... 
    // Matches if maybe is not null, and then against the inner pattern "var s"
  }
  ```  
- [**Object patterns**][object]: the match exposes getters on the object type.
  ```dart
  switch (Shape shape) {
    case Rect(width: var w, height: var h): ...
    // Matches if shape is of type Rect, and then against the properties of Rect
  }
  ```  

> Values destructured by a pattern in a case become local variables,

Reviewer 3: I'm not sure if it makes a lot of sense to list out all of the kinds of patterns here when they are also in the other reference page. Also, most of these patterns aren't specific to switch cases, so it arguably makes as much sense to have them here as it would to have them under any of the preceding sections: you can have object, list, and map patterns in pattern variable declarations and pattern assignments too.

Author: Ok, good point. I placed this section here (originally from the switch case content) before writing the patterns reference page, but now that that exists this can definitely go

only available in the body of that case. 

### Object destructuring

[Object patterns][object] match against named object types, allowing

> you to destructure their data using the getters the object's class already exposes.

Reviewer 4: nit: My understanding is an object's properties (fields or getters) can be used:
```suggestion
you to destructure their data using the object's properties.
```

For example, you can use object destructuring in a [for-in loop][for] to iterate-over
and destructure the [`MapEntry`][] objects that a `<Map>.entries` call
returns:

```dart
Map<String, int> hist = // ...

for (var MapEntry(key: key, value: count) in hist.entries) {
  print('$key occurred $count times');
}
```

The object pattern checks that `hist.entries` has the named type `MapEntry`,
and then recurses into the named field subpatterns `key` and `value`.
It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
and binds the results to local variables `key` and `count`, respectively.

Binding the result of a getter call to a variable of the same name is a common
use case, so object patterns can also infer the getter name from the
[variable subpattern][variable]. This allows you to simplify the variable pattern
from something redundant like `key: key` to just `:key`:

```dart
for (var MapEntry(:key, value: count) in hist.entries) { 
  print('$key occurred $count times');
}

> ```

Reviewer 3: This section feels a little tacked on because it's more or less just a description of how object patterns work in general, so it overlaps the previous object pattern bullet point and the section on object patterns in the reference. Maybe expand/merge this into section that talks about using patterns in for/for-in loops (which isn't otherwise shown on this doc) and make this an example under that section?

> ### Algebraic datatypes 

Reviewer 3: If you don't want to move this as an example under the switch cases section, then I think it would help to separate out the "places you can use patterns" and "idioms patterns are helpful for" sets of sections under different subheadings like:

```
## Places patterns can appear

### Pattern variable declarations

...

### Pattern assignments

...

### Switch cases

...

## Use cases for patterns

### Multiple returns

...

### Algebraic datatype-style code

...

### Validating JSON

...

```

Author: I compared your two new structuring ideas ([the other one](https://github.com/dart-lang/site-www/pull/4810/files#r1183153697)) and I think this is the way to go.

Object destructuring and switch cases are conducive to writing
code in an [algebraic datatype][] style: 
when you have a family of related types and an operation that needs
specific behavior for each type.

Instead of implementing the operation as an instance method for every respective type,
you can keep the operations in a single function, switching over subtypes for
variations in behavior:

```dart
abstract class Shape {
  double calculateArea();
}

class Square implements Shape {
  final double length;
  Square(this.length);

  double calculateArea() => length * length;
}

class Circle implements Shape {
  final double radius;
  Circle(this.radius);

  double calculateArea() => math.pi * radius * radius;
}

> double calculateArea(Shape shape) =>

Reviewer: This code is an error as written, since switch expressions must be exhaustive.  You could make `Shape` be `sealed`, or you could use a switch statement, or you could add a default case.  Given that you are talking about algebraic datatypes here, I think the first solution is the correct one, since that's the use case that `sealed` addresses.

  switch (shape) {
    Square(length: var l) => l * l,
    Circle(radius: var r) => math.pi * r * r
  };
```

### Validating incoming JSON

[Map][] and [list][] patterns work well for destructuring key-value pairs in
JSON data:

```dart 
var json = {'user': ['Lily', 13]};
var {'user': [name, age]} = json;
```

The example above makes sense when you already know that the JSON
data has the structure you expect.
If the data comes from an external source, like over the network as is commonly
the case, you can't be sure of it's structure and
should validate it first. 

Without patterns, validation is verbose:

```dart
if (json is Map<String, dynamic> &&
    json.length == 1 && json.containsKey('user')) {
  var user = json[user];
  if (user is List<dynamic> && user.length == 2 &&
      user[0] is String && user[1] is int) {
    var name = user[2] as String;
    var age = user[1] as int;
    print('User $name is $age years old.');
  }
}
```

A single [case pattern](#switch-cases) in a switch achieves the same validation

> as above, but is much clearer, more declarative, and much shorter:

Reviewer 3: This might be a good place to use `if-case` instead of a switch since there's only a single case.

Author: Changed it to:

> A single [case pattern][] (conducive to an [if-case][] statement)
> achieves the same validation as above, but is much clearer, more declarative,
> and much shorter:
> 
> ```dart
> if (json case {'user': [String name, int age]}) 
>   print('User $name is $age years old.');
> ```

The parenthetical about if-case, keep or toss?

---
title: Patterns
description: Summary of patterns in Dart.
---

{{site.alert.version-note}}
  Patterns require a [language version][] of at least 3.0.
{{site.alert.end}}

Patterns are a syntactic category in the Dart language, like statements and expressions.
A pattern represents the shape of a set of values that it may match against actual
values.

This page describes:
- What patterns do
- Where patterns are allowed in Dart code

> - Common use cases for patterns

Author: Can I say "What the common use cases for patterns are"? Or is that grammatically worse? "What are..." sounds weird to me but I do want to think of something consistent for this point

> - Common use cases for patterns

Reviewer 5: **comment:** Oh, yeah. It reads like a question how I wrote it. Your revision would be better. You can edit my comment suggestion to your wording and accept _that_.

To learn about the different kinds of patterns, visit the [pattern types][types]
page.

## What patterns do

In general, a pattern may **match** a value, **destructure** a value, or both,
depending on the context and shape of the pattern.

First, given a value, pattern matching allows you to check whether the value has
a certain shape, or is a certain constant, or is equal to something else, or has
a certain type.

Then, destructuring provides you with a convenient declarative syntax to break
that value into its constituent parts, possibly binding variables to some or all
of those parts in the process.

### Matching

A pattern is always tested against a value to determine if the value has the form

> you expect, or in other words, if the value _matches_ the pattern. 

Author: The Dart team has been pretty clear against using "Dart" as an actor for anything in the docs, and I agree. I'll take the rest of the notes and say "Patterns compare against..." or something like that.

What constitutes a match depends on [what kind of pattern][types] is being used.
For example, a constant pattern matches if the value is equal to the pattern's 
constant:

```dart
switch (number) {
  // Constant pattern matches if 1 == number.
  case 1: // ...        
}
```

Many patterns make use of subpatterns (sometimes called outer and inner patterns,
respectively). Patterns match recursively on their subpatterns.
For example, the individual fields of any [collection-type][] pattern could be 
[variable patterns][variable] or [constant patterns][constant]:

```dart
switch (obj) {
  // List pattern [a,b] matches obj first if obj is a list wth two fields,
  // then if its fields match the constant subpatterns 'a' and 'b'.
  case [a, b]: // ...   
}                    
```

Patterns can ignore parts of a matched value by using a [wildcard pattern][]

> as a place holder, or a [rest element][] in the case of list patterns.

Author: I'm going to use "you" as the actor here, instead of "patterns", as it is a personal choice. But incorporating the gist of it.

### Destructuring

When an object and pattern match, the pattern can then access the object's data 
and extract it in parts. In other words, the pattern _destructures_ the object:

```dart
var numList = [1, 2, 3];
// List pattern [a, b, c] destructures the three elements from numList...
var [a, b, c] = numList;
// ...and assigns them to new variables.
print(a + b + c);          
```

You can nest [any kind of pattern][types] inside a destructuring pattern, 
For example, this case pattern matches and destructures a two-element
list whose first element is `a` or `b`:

```dart
switch (list) {
  case ['a' || 'b', var c]: 
  // ...
}
```

## Places patterns can appear

You can use patterns in several places in the Dart language:

<a id="pattern-uses"></a>

- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
- [for and for-in loops][for]
- [if-case][if] and [switch-case][switch]
- Control flow in [collection literals][]

This section describes common use cases for matching and destructuring with patterns, 

> combining various contexts, pattern types, and surrounding code constructs.

Author: Just got rid of the second clause, it doesn't really add anything

### Variable declaration

Anywhere Dart allows local variable declaration, like in the body of a function,
you can also use a _pattern variable declaration_. The pattern matches against
and destructures the value on the right of the declaration, binding it to
new local variables:

```dart
// Declares new variables a, b, and c.
var (a, [b, c]) = ("str", [1, 2]);
```

A pattern variable declaration must start with either `var` or `final`, followed
by a pattern. 

### Variable assignment 

A _variable assignment pattern_ on the left side of an assignment destructures the
matched object and assigns the values to _existing_ variables,
instead of binding new ones. 

One convenient use case is swapping the values of two variables without needing
to declare a third temporary one:

```dart
var (a, b) = ('left', 'right');
(b, a) = (a, b); // Swap.
print('$a $b'); // Prints "right left".
```

### Switch statements and expressions

Every case clause contains a pattern. This applies to [switch statements][switch]
and [expressions][], as well as [if-case statements][if].

> _Case patterns_ are [refutable][];
> they allow control flow to either match and destructure the object being switched
> on, or continue execution if the object doesn't match. Case patterns can be
> [any kind of pattern][types]:

Reviewer 5: **issue:** If you can use a semi-colon, you can use a period.
```suggestion
_Case patterns_ are [refutable][].
They control flow toward one of two outcomes:
* Match and destructure the switched object.
* Continue execution if the object doesn't match.
Case patterns can be [any kind of pattern][types].
```

**nit:** I would change the first sentence to active voice, but that would involve changing the link.

Author: I'm not sure how the first sentence can be made active. Ignoring the link, what would you do?

Reviewer 5: **comment:** What refutes case patterns? That would explain it.

That said, I'm trying to soften my approach to passive voice. The **When do I use passive voice?** section of [this page](https://advice.writing.utoronto.ca/revising/passive-voice/) from the University of Toronto seems to be a good guide. Look at that to see if what you have here just makes more sense to keep as is.

Author: Oh, ok. It didn't even cross my mind that that kind of sentence structure (Case patterns are refutable) is passive voice, good to know!

Who or what is refuting them is really hard to say so I'm going to leave this for now. I think it might be the pattern that does the refuting, but also is the one being refuted...idk. I vaguely remember Bob saying "refutable/irrefutable" are _types_ of patterns, so the passive voice is built into the descriptors.

```dart
switch (obj) {
  case 1: ...
    // Matches if 1 == obj.
  case >= first && <= last: ...  
    // Matches if the value of obj is between the constant values of 'first' and 'last'.
  case (var a, var b): ...
    // Matches if obj is a record with two fields, then assigns the fields to 'a' and 'b'.
  case Rect(width: var w, height: var h): ...
    // Matches if obj is an instance of Rect, and then against the properties of Rect.
  default: ...
}
```

> Values destructured by a pattern in a case become local variables,

Reviewer 5: **issue:** Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
Values that a pattern destructures in a case become local variables.
```

> only available in the body of that case. 

Reviewer 5: **issue:** Rephrase.
```suggestion
This limits the scope of these local variables to the body of that `case` statement. 
```

Author: Dropped "statement" at the end because want it to be generalized for expressions too

### For and for-in loops

You can use patterns in [for and for-in loops][for] to iterate-over and destructure
values in a collection.

This example uses [object destructuring][object] in a for-in loop to destructure
the [`MapEntry`][] objects that a `<Map>.entries` call returns:

```dart
Map<String, int> hist = // ...

for (var MapEntry(key: key, value: count) in hist.entries) {
  print('$key occurred $count times');
}
```

The object pattern checks that `hist.entries` has the named type `MapEntry`,
and then recurses into the named field subpatterns `key` and `value`.
It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
and binds the results to local variables `key` and `count`, respectively.

Binding the result of a getter call to a variable of the same name is a common
use case, so object patterns can also infer the getter name from the
[variable subpattern][variable]. This allows you to simplify the variable pattern
from something redundant like `key: key` to just `:key`:

```dart
for (var MapEntry(:key, value: count) in hist.entries) { 
  print('$key occurred $count times');
}
```

## Uses cases for patterns

> The [section above](#places-patterns-can-appear)

Reviewer 5: **issue:** Don't use physical directionality in documentation per [GSG](https://developers.google.com/style/word-list#above).

> describes _how_ you can use patterns in Dart code constructs, 
> sometimes with interesting uses cases as examples (like [swapping](#variable-assignment)
> the values of two variables, or
> [destructuring key-value pairs](#for-and-for-in-loops)
> in a map). 

Reviewer 5: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
describes _how_ you can use patterns in Dart code constructs. 
Dart code constructs have interesting uses cases as examples.
These include [swapping](#variable-assignment)
the values of two variables
and [destructuring key-value pairs](#for-and-for-in-loops)
in a map. 
```

> This section describes even more use cases, answering _when and why_ you might
> want to use patterns, the kinds of problems they solve, and the idioms they're 
> best suited for.

Reviewer 5: **issue:** 
Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
This section describes even more use cases that answer:
* _when and why_ you might want to use patterns.
* the kinds of problems they solve.
* the idioms for which they're best suited.
```

### Destructuring multiple returns

[Records][] allow aggregating and returning multiple values from a single function
call. Patterns add the ability to destructure a record's fields
directly into local variables, inline with the function call.

Instead of individually declaring new local variables for each record field,
like this:

```dart
main() {
  var info = userInfo(json);
  var name = info.$1;
  var age = info.$2;
}
```

You can destructure the fields of a record that a function returns into local
variables using a [variable declaration](#variable-declaration) or
[assigment pattern](#variable-assignment), and a record pattern as its subpattern:

```dart
main() {
  var (name, age) = userInfo(json);
}
```

### Destructuring class instances

[Object patterns][object] match against named object types, allowing
you to destructure their data using the getters the object's class already exposes.

To destructure an instance of a class, use the named type, followed by the properties
to destructure enclosed in parenthesis:

```dart
main() {
  final Foo myFoo = Foo(one: 'one', two: 2);
  var Foo(:one, :two) = myFoo;
  print('one $one, two $two');
}
```

### Algebraic datatypes 

> Object destructuring and switch cases are conducive to writing
> code in an [algebraic datatype][] style: 
> when you have a family of related types and an operation that needs
> specific behavior for each type.

Reviewer 5: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading). This is a 33 word sentence.
```suggestion
Object destructuring and switch cases aid in writing
code in an [algebraic datatype][] style.
Use these methods when you have:
* a family of related types
* an operation that needs specific behavior for each type
```

> Instead of implementing the operation as an instance method for every respective type,
> you can keep the operations in a single function, switching over subtypes for
> variations in behavior:

Reviewer 5: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading). One 29 word sentence to three ~10 word sentences.
```suggestion
This provides an alternative to each type having its own instance method.
Keep the operations, or instance methods, in one function.
Use a switch statement to to through subtypes for behavior variations.
```

Author: I liked the way this sentence flowed, so I cut some words out to get it down to 26 per GSG instead of restructuring it 😬 

> Instead of implementing the operation as an instance method for every type,
keep the operations in a single function, switching over subtypes for
behavior variations:

```dart
sealed class Shape {}

class Square implements Shape {
  final double length;
  Square(this.length);
}

class Circle implements Shape {
  final double radius;
  Circle(this.radius);
}

double calculateArea(Shape shape) =>
  switch (shape) {
    Square(length: var l) => l * l,
    Circle(radius: var r) => math.pi * r * r
  };
```

### Validating incoming JSON

[Map][] and [list][] patterns work well for destructuring key-value pairs in
JSON data:

```dart 
var json = {'user': ['Lily', 13]};
var {'user': [name, age]} = json;
```

> The example above makes sense when you already know that the JSON
> data has the structure you expect.

Reviewer 5: **nits**: Don't use physical direction indicators. Also, if you know something, the already would be implied. 
```suggestion
When you know that the JSON data has the structure you expect,
the previous example makes sense.
```

> If the data comes from an external source, like over the network as is commonly
> the case, you can't be sure of it's structure and
> should validate it first. 

Reviewer 5: **issues:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
**typo:** `its` not `it's`.
```suggestion
If the data comes from an external source, you can't validate its structure.
You should validate it first. 
```

Without patterns, validation is verbose:

```dart
if (json is Map<String, dynamic> &&
    json.length == 1 && json.containsKey('user')) {
  var user = json[user];
  if (user is List<dynamic> && user.length == 2 &&
      user[0] is String && user[1] is int) {
    var name = user[2] as String;
    var age = user[1] as int;
    print('User $name is $age years old.');
  }
}
```

> A single [case pattern](#switch-statements-and-expressions)
> (conducive to an [if-case][if] statement)
> achieves the same validation as above, but is much clearer, more declarative,
> and much shorter:

Reviewer 5: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
One [case pattern](#switch-statements-and-expressions) validates
in the same way as the previous example did.
It does so with more clarity, more declaration, and less size.
```

```dart
if (json case {'user': [String name, int age]}) 
  print('User $name is $age years old.');
```

This case pattern simultaneously validates that:

> - `json` is a map (because it must first match the outer [map pattern][map] to proceed).

Reviewer 5: **issue:** Remove the parenthetical.
```suggestion
- `json` is a map because it must first match the outer [map pattern][map] to proceed.
```

Every case clause contains a pattern. This applies to [switch statements][switch]
and [expressions][], as well as [if-case statements][if].
You can use [any kind of pattern][types] in a case.

_Case patterns_ are [refutable][].
They allow control flow to either:

> - Match and destructure the object being switched on,
> - Or continue execution if the object doesn't match.

Reviewer 5: **nit:** For parallelism in the list, change this slightly:
```suggestion
- Match and destructure the object being switched on, or
- Continue execution if the object doesn't match.
```

> - Or continue execution if the object doesn't match.

Author: I guess the "or" isn't really even necessary

## Uses cases for patterns

The [previous section](#places-patterns-can-appear)
describes _how_ patterns fit into other Dart code constructs. 
You saw some interesting uses cases as examples, like [swapping](#variable-assignment)
the values of two variables, or
[destructuring key-value pairs](#for-and-for-in-loops)
in a map. This section describes even more use cases, answering:

- _When and why_ you might want to use patterns.

> - The kinds of problems they solve.
> - The idioms for which they're best suited.

Reviewer 5: **suggestion:** Parallelism.
```suggestion
- _What_ kinds of problems they solve.
- _Which_ idioms they best suit.
```

var {'user': [name, age]} = json;
```

> IF you know that the JSON data has the structure you expect,

Reviewer 5: **typo:**
```suggestion
If you know that the JSON data has the structure you expect,
```

the previous example is realistic.

> But data typically comes from an external source, like over the network.

Reviewer 5: **nit:** Rephrase.
```suggestion
But the typical data source comes from an external origin, like over the network.
```

```

A single [case pattern](#switch-statements-and-expressions)
can achieve the same validation.

> Single cases are conducive to [if-case][if] statements.

Reviewer 5: **nit:** This reads with some complexity. Simplify the word choice.
```suggestion
Single cases work best as [if-case][if] statements.
```
